File: 40-combination-sum-ii/combination-sum-ii.java
class Solution {
    //THIS IS THE METHOD FOR GETTING NO DUPLICATES WITHOUT HAVING A SET....
    void check(List<Integer> s,List<List<Integer>> ans,int i,int[] candidates,int target){
        if(target==0){
            ans.add(new ArrayList<>(s));
            return ;
        }
        for(int j=i;j<candidates.length;j++){
            if(j==i || candidates[j]!=candidates[j-1]){
               if(candidates[j]<=target){
                target=target-candidates[j];
                s.add(candidates[j]);
                check(s,ans,j+1,candidates,target);
                target=target+candidates[j];
                s.remove(s.size()-1);            
               }
               else
                 return;
            }
           
        }
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> ans=new ArrayList<>();
        List<Integer> s=new ArrayList<>();

        check(s,ans,0,candidates,target);
        return ans;
    }
}

File: 40-combination-sum-ii/combination-sum-ii.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """

        def find(i, ans, subans, tot):
            if tot == target:
                temp = [k for k in subans]
                ans.append(temp)
                print(ans)
                return 
            if tot > target :
                return
            
            #pick 
            for j in range(i, len(candidates)):
                if j == i or candidates[j - 1] != candidates[j]:
                    if tot + candidates[j] <= target:
                        subans.append(candidates[j])
                        tot += candidates[j]
                        find(j + 1, ans, subans, tot)
                        tot -= candidates[j]
                        subans.pop(-1)
                    else:
                        return
            
            return

                
        
        candidates.sort()
        ans = []
        find(0, ans, [], 0)
        return ans